{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cogent3.core.annotation_db import GffAnnotationDb\n",
    "from cogent3.core.annotation_db import GenbankAnnotationDb\n",
    "from cogent3.parse.gff import gff_parser\n",
    "from cogent3 import open_\n",
    "from cogent3.parse.genbank import MinimalGenbankParser\n",
    "from cogent3 import load_unaligned_seqs\n",
    "gff_path = \"/Users/kiratalreja/Downloads/prok_NoLocusTags.gff\"\n",
    "genbank_path = \"/Users/kiratalreja/Downloads/NC_000913.3.gb\"\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_make_db():\n",
    "    \"\"\"Test if the gff3 database is correctly created\n",
    "    in terms of the names of columns & the number of entries\"\"\"\n",
    "\n",
    "    db = GffAnnotationDb(gff_path)\n",
    "\n",
    "    \"\"\"Test the name of columns created\"\"\"\n",
    "    sql_columns = []\n",
    "    data = db.db.execute(\"\"\"SELECT * FROM GFF\"\"\")\n",
    "    for column in data.description:\n",
    "        sql_columns.append(column[0])\n",
    "    got = sql_columns\n",
    "    parsed_gff = gff_parser(gff_path)\n",
    "    expected = list((list(parsed_gff)[0]).keys())[:-1]\n",
    "    assert got == expected\n",
    "    \n",
    "    sql_columns = []\n",
    "    db = GenbankAnnotationDb(genbank_path)\n",
    "    data = db.db.execute(\"\"\"SELECT * FROM GENBANK\"\"\")\n",
    "    for column in data.description:\n",
    "        sql_columns.append(column[0])\n",
    "    got = sql_columns\n",
    "    expected = [\"LocusID\",\"Type\",\"Spans\",\"Locus_Tag\",\"Start\",\"End\",\"Strand\"]\n",
    "    assert got == expected\n",
    "    \n",
    "   \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_make_sql_query():\n",
    "    \"\"\"Test that the SQLlite queries are correctly formed\"\"\"\n",
    "    \n",
    "    fasta_path = \"/Users/kiratalreja/Desktop/short.fa\"\n",
    "    seqs = load_unaligned_seqs(fasta_path, moltype=\"dna\")\n",
    "    seq = seqs.seqs[0]\n",
    "\n",
    "    gff_path = \"/Users/kiratalreja/Downloads/prok_NoLocusTags.gff\"\n",
    "    db = GffAnnotationDb(gff_path)\n",
    "\n",
    "    \"\"\"only the bio_type is provided\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE Type == ?', ['gene'])\n",
    "    got = db._make_sql_query(seq_name=None,bio_type=\"gene\",identifier=None,start=None,end=None)\n",
    "    assert got == expected \n",
    "\n",
    "    \"\"\"only the identifier is provided\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE Attributes like ?', ['%RandomIdentifier%'])\n",
    "    got = db._make_sql_query(seq_name=None,bio_type=None,identifier=\"RandomIdentifier\",start=None,end=None)\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"both identifier and bio_type provided\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE Type == ? AND Attributes like ?',['CDS', '%RandomIdentifier%'])\n",
    "    got = db._make_sql_query(seq_name=None,bio_type=\"CDS\",identifier=\"RandomIdentifier\",start=None,end=None)\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"start provided along with identifier and bio_type\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE Type == ? AND Attributes like ? AND Start >= ?',['CDS', '%RandomIdentifier%', 0])\n",
    "    got = db._make_sql_query(seq_name=None,bio_type=\"CDS\",identifier=\"RandomIdentifier\",start=0,end=None)\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"end provided along with identifier and bio_type\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE Type == ? AND Attributes like ? AND End < ?',['CDS', '%RandomIdentifier%', 5000])\n",
    "    got = db._make_sql_query(seq_name=None,bio_type=\"CDS\",identifier=\"RandomIdentifier\",start=None,end=5000)\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"start and end provided along with identifier and bio_type\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE Type == ? AND Attributes like ? AND Start >= ? AND End < ?',['CDS', '%RandomIdentifier%', 0, 5000])\n",
    "    got = db._make_sql_query(seq_name=None,bio_type=\"CDS\",identifier=\"RandomIdentifier\",start=0,end=5000)\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"all five attributes provided\"\"\"\n",
    "    expected = ('SELECT * FROM GFF WHERE SeqID == ? AND Type == ? AND Attributes like ? AND Start >= ? AND End < ?', ['1', 'CDS', '%RandomIdentifier%', 0, 5000])\n",
    "    got = db._make_sql_query(seq_name=\"1\",bio_type=\"CDS\",identifier=\"RandomIdentifier\",start=0,end=5000)\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"check exception when both bio_type and identifier are missing\"\"\"\n",
    "    import pytest\n",
    "\n",
    "    with pytest.raises(ValueError):\n",
    "        db._make_sql_query(seq_name=None,bio_type=None,identifier=None,start=None,end=None)\n",
    "\n",
    "    \"\"\"check exception when both bio_type and identifier are missing, even if other attributes\"\"\"\n",
    "\n",
    "    with pytest.raises(ValueError):\n",
    "        db._make_sql_query(seq_name=\"1\",bio_type=None,identifier=None,start=0,end=1000)\n",
    "\n",
    "    \"\"\"check exception when only seq_name is provided\"\"\"\n",
    "\n",
    "    with pytest.raises(ValueError):\n",
    "        db._make_sql_query(seq_name=\"1\",bio_type=None,identifier=None,start=None,end=None)\n",
    "\n",
    "    \"\"\"check exception when only start is provided\"\"\"\n",
    "\n",
    "    with pytest.raises(ValueError):\n",
    "        db._make_sql_query(seq_name=None,bio_type=None,identifier=None,start=0,end=None)\n",
    "    \n",
    "    \"\"\"check exception when only end is provided\"\"\"\n",
    "\n",
    "    with pytest.raises(ValueError):\n",
    "        db._make_sql_query(seq_name=None,bio_type=None,identifier=None,start=None,end=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_populate_from_file():\n",
    "    \"\"\"Test that the database is populated with the correct\n",
    "    number of columns\"\"\"\n",
    "    from cogent3.core.annotation import GffAnnotationDb\n",
    "    from cogent3.parse.gff import gff_parser\n",
    "\n",
    "    gff_path = \"/Users/kiratalreja/Downloads/prok_NoLocusTags.gff\"\n",
    "    db = GffAnnotationDb(gff_path)\n",
    "\n",
    "    \"\"\"test the number of rows populated\"\"\"\n",
    "    db.db.execute(\"\"\" SELECT * FROM GFF \"\"\")\n",
    "    got = len(list(db.db.fetchall()))\n",
    "    parsed_gff = gff_parser(gff_path)\n",
    "    expected = len(list(parsed_gff))\n",
    "    assert got == expected\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_db_query():\n",
    "\n",
    "    \"\"\"Test that the SQL query returns the correct\n",
    "    number of rows for different combinations of bio_type/identifier\"\"\"\n",
    "\n",
    "    from cogent3 import load_unaligned_seqs\n",
    "    from cogent3.core.annotation import GffAnnotationDb\n",
    "\n",
    "    fasta_path = \"/Users/kiratalreja/Desktop/short.fa\"\n",
    "    seqs = load_unaligned_seqs(fasta_path, moltype=\"dna\")\n",
    "    seq = seqs.seqs[0]\n",
    "\n",
    "    gff_path = \"/Users/kiratalreja/Downloads/prok_NoLocusTags.gff\"\n",
    "    db = GffAnnotationDb()\n",
    "    db.populate_from_file(gff_path)\n",
    "\n",
    "    \"\"\"multiple hits for the same identifier\"\"\"\n",
    "    got = len(db.db_query(start=0, end=len(seq), identifier=\"CDS4\"))\n",
    "    expected = 2\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"query for an ID and recieve the children to the ID along with the parent\"\"\"\n",
    "    got = len(db.db_query(start=0, end=len(seq), identifier=\"gene4\"))\n",
    "    expected = 3\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"query for an ID, with no children\"\"\"\n",
    "    got = len(db.db_query(start=0, end=len(seq), identifier=\"trna1\"))\n",
    "    expected = 1\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"query for a bio type, with multiple hits\"\"\"\n",
    "    got = len(db.db_query(start=0, end=len(seq), bio_type=\"gene\"))\n",
    "    expected = 8\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"query for an ID & a bio type, with a single hit\"\"\"\n",
    "    got = len(db.db_query(start=0, end=len(seq), identifier=\"gene0\", bio_type=\"CDS\"))\n",
    "    expected = 1\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"query for an ID & a bio type, with multiple hits\"\"\"\n",
    "    got = len(db.db_query(start=0, end=len(seq), identifier=\"CDS4\", bio_type=\"CDS\"))\n",
    "    expected = 2\n",
    "    assert got == expected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_find_records():\n",
    "\n",
    "    \"\"\"Test that the coordinates the grouped correctly, and features\n",
    "    formed properly\"\"\"\n",
    "\n",
    "    from cogent3 import load_unaligned_seqs\n",
    "    from cogent3.core.annotation import GffAnnotationDb\n",
    "\n",
    "    gff_path = \"/Users/kiratalreja/Downloads/prok_NoLocusTags.gff\"\n",
    "    db = GffAnnotationDb()\n",
    "    db.populate_from_file(gff_path)\n",
    "    fasta_path = \"/Users/kiratalreja/Desktop/short.fa\"\n",
    "    seqs = load_unaligned_seqs(fasta_path, moltype=\"dna\")\n",
    "    seq = seqs.seqs[0]\n",
    "\n",
    "    \"\"\"combine rows with the same ID\"\"\"\n",
    "    got = len(db.find_records(start=0, end=len(seq), identifier=\"CDS4\"))\n",
    "    expected = 1\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"combine rows with the same ID, when bio_type given\"\"\"\n",
    "    got = len(db.find_records(start=0, end=len(seq), bio_type=\"CDS\"))\n",
    "    expected = 6\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"combine rows with the same ID, when children rows are fetched with the parent\"\"\"\n",
    "    got = len(db.find_records(start=0, end=len(seq), identifier=\"gene4\"))\n",
    "    expected = 2\n",
    "    assert got == expected\n",
    "\n",
    "    \"\"\"unique ID, single row returned\"\"\"\n",
    "    got = len(db.find_records(start=0, end=len(seq), identifier=\"id020000\"))\n",
    "    expected = 1\n",
    "    assert got == expected\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_find_records()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.0 ('c310dev')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "cdab2fb02dc4d68ed9e9d327e285305822cd116f79900361a665d6985eb56184"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
